src/Temporalio/Bridge/CustomSlotSupplier.cs

```diff
@@ -0,0 +1,179 @@
+            ILoggerFactory loggerFactory)
+        {
+            this.logger = loggerFactory.CreateLogger<CustomSlotSupplier>();
+            this.userSupplier = userSupplier;
+
+            var interopCallbacks = new Interop.CustomSlotSupplierCallbacks
+            {
+                reserve = FunctionPointer<Interop.CustomReserveSlotCallback>(Reserve),
+                cancel_reserve = FunctionPointer<Interop.CustomCancelReserveCallback>(CancelReserve),
+                try_reserve = FunctionPointer<Interop.CustomTryReserveSlotCallback>(TryReserve),
+                mark_used = FunctionPointer<Interop.CustomMarkSlotUsedCallback>(MarkUsed),
+                release = FunctionPointer<Interop.CustomReleaseSlotCallback>(Release),
+            };
+
+            PinCallbackHolder(interopCallbacks);
+        }
+
+        private static void SetCancelTokenOnCtx(ref SlotReserveCtx ctx, CancellationTokenSource cancelTokenSrc)
+        {
+            unsafe
+            {
+                try
+                {
+                    var handle = GCHandle.Alloc(cancelTokenSrc);
+                    fixed (Interop.SlotReserveCtx* p = &ctx)
+                    {
+                        Interop.Methods.set_reserve_cancel_target(p, GCHandle.ToIntPtr(handle).ToPointer());
+                    }
+                }
+                catch (Exception e)
+                {
+                    Console.WriteLine($"Error setting cancel token on ctx: {e}");
+                    throw;
+                }
+            }
+        }
+
+        private unsafe void Reserve(Interop.SlotReserveCtx ctx, void* sender)
+        {
+            SafeReserve(ctx, new IntPtr(sender));
+        }
+
+        private unsafe void CancelReserve(void* tokenSrc)
+        {
+            var handle = GCHandle.FromIntPtr(new IntPtr(tokenSrc));
+            var cancelTokenSrc = (CancellationTokenSource)handle.Target!;
+            cancelTokenSrc.Cancel();
+            handle.Free();
+        }
+
+        private void SafeReserve(Interop.SlotReserveCtx ctx, IntPtr sender)
+        {
+            var reserveTask = Task.Run(async () =>
+            {
+                var cancelTokenSrc = new System.Threading.CancellationTokenSource();
+                SetCancelTokenOnCtx(ref ctx, cancelTokenSrc);
+                while (true)
+                {
+                    try
+                    {
+                        var permit = await userSupplier.ReserveSlotAsync(
+                            new(ctx), cancelTokenSrc.Token).ConfigureAwait(false);
+                        var usedPermitId = AddPermitToMap(permit);
+                        unsafe
+                        {
+                            Interop.Methods.complete_async_reserve(sender.ToPointer(), new(usedPermitId));
+                        }
+                        cancelTokenSrc.Dispose();
+                        return;
+                    }
+                    catch (OperationCanceledException)
```

Comment:
```suggestion
                    catch (OperationCanceledException) when (cancelTokenSrc.IsCancellationRequested)
```

I assume there will be a missed reservation if a user raised this exception for whatever reason unrelated to our cancellation. In fact, should we either 1) _always_ call `complete_async_reserve` before returning, even on cancel, and/or 2) _always_ avoid calling `complete_async_reserve` if cancellation is requested (user can choose to ignore cancellation token)

Reply:
It won't matter - because if the token is cancelled Rust has already given up waiting for anything to come back through the completion channel, so if something is sent or not, either way it's ignored.

Reply:
I am saying `OperationCanceledException` can be thrown from `ReserveSlotAsync` without it being the result of _this_ cancellation token, hence the `when` suggestion here.

But in addition to that, remember that `Box::into_raw(Box::new(tx))` means that Rust will never drop the tx side of the oneshot until it's `Box::from_raw`'d in `complete_async_reserve`. So to not have memory leaks, you probably need to call `complete_async_reserve` even on cancel, but just give it a user permit of 0 or something.

Reply:
Ah, yeah I get what you mean now